skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,434 @@
+ * <p>
+ *  More information at Skeletonize3D homepage:
+ *  http://fiji.sc/Skeletonize3D
+ *
+ * @version 1.0 11/13/2015 (unique BSD licensed version for scikit-image)
+ * @author Ignacio Arganda-Carreras (iargandacarreras at gmail.com)
+
+Porting to Cython was done by Evgeni Burovski (evgeny.burovskiy@gmail.com).
+
+References
+----------
+
+.. [Lee94] T.-C. Lee, R.L. Kashyap and C.-N. Chu, Building skeleton models
+       via 3-D medial surface/axis thinning algorithms.
+       Computer Vision, Graphics, and Image Processing, 56(6):462-478, 1994.
+
+
+.. [IAC15] Ignacio Arganda-Carreras, 2015. Skeletonize3D plugin for ImageJ(C).
+           http://fiji.sc/Skeletonize3D
+
+"""
+from __future__ import division, print_function, absolute_import
+
+from libc.string cimport memcpy
+
+import numpy as np
+from numpy cimport npy_intp, npy_uint8, ndarray
+cimport cython
+
+ctypedef npy_uint8 pixel_type
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _compute_thin_image(pixel_type[:, :, ::1] img not None):
+    """Compute a thin image.
+
+    Loop through the image multiple times, removing "simple" points, i.e.
+    those point which can be removed without changing local connectivity in the
+    3x3x3 neighborhood of a point.
+
+    This routine implements the two-pass algorthim of [Lee94]_. Namely,
+    for each of the six border types (positive and negative x-, y- and z-),
+    the algorithm first collects all possibly deletable points, and then
+    performs a sequential rechecking.
+
+    The input, `img`, is assumed to be a 3D binary image in the
+    (p, r, c) format [i.e., C ordered array], filled by zeros (background) and
+    ones. Furthermore, `img` is assumed to be padded by zeros from all
+    directions --- this way the zero boundary conditions are authomatic
+    and there is need to guard against out-of-bounds access.
+
+    """
+    cdef:
+        int unchanged_borders = 0, curr_border, num_borders
+        int borders[6]
+        npy_intp p, r, c
+        bint no_change
+
+        list simple_border_points
+        Py_ssize_t num_simple_points, i, j
+        (npy_intp, npy_intp, npy_intp) point
+
+        pixel_type neighb[27]
+    borders[:] = [4, 3, 2, 1, 5, 6]
+
+    # no need to worry about the z direction if the original image is 2D.
+    if img.shape[0] == 3:
+        num_borders = 4
+    else:
+        num_borders = 6
+
+    # loop through the image several times until there is no change for all
+    # the six border types
+    while unchanged_borders < num_borders:
+        unchanged_borders = 0
+        for j in range(num_borders):
+            curr_border = borders[j]
+
+            simple_border_points = _loop_through(img, curr_border)
+
+            # sequential re-checking to preserve connectivity when deleting
+            # in a parallel way
+            no_change = True
+            num_border_points = len(simple_border_points)
```

Comment:
That one is interesting. Prompted by this comment I wondered why `cython -a` did not show a bright yellow line here. It turns out that at least cython 0.23.4 manages to infer the type for `num_border_points` here:

```
     __pyx_t_6 = PyList_GET_SIZE(__pyx_v_simple_border_points); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_v_num_border_points = __pyx_t_6;
```

Nonetheless, it was a typo (`num_border_points` vs `num_simple_points`) which is now corrected. Thanks!